### DISCUSSION TITLE
Implement state preparation algorithm

### DISCUSSION THREAD

**Discussion by hros** [QuantumBFS/Yao.jl#472] [General]
Hi,
I want to implement the algorithm from "[An Efficient Algorithm for Sparse Quantum State Preparation](http://ww.unixer.de/publications/img/quantum_dac.pdf)" in Yao.
I have (quite) a few questions:
1. what is the recommended way to create a sparse superposition?
I could do `r = 0.6 product_state(bit"111111110001100") + 0.8 product_state(bit"111111110001010")`
 a) Is it stored efficiently?
 b) can I easily retrieve the states? if I use `statevec`, I get a vector whose size is 2^num_qubits rather than the number of states
This is especially important to this specific algorithm, as it accepts a very large, yet sparse, superposition and iteratively reduces the number of states by applying circuits. Thus, I need to read the individual states to build the circuit and then read the states of the resulting operation until reaching the final state (not a superposition).
2. I saw many references to `active` and `inactive` qubits in Yao, but I did not find an explanation for what that means
3. I used `matblock` to create a new unitary. is it automatically "reversed" with the adjoint `'` operator
4. I would like to export and import circuits in QASM format. Does Yao support that? I found a repository, [YaoQasm](https://github.com/QuantumBFS/YaoQASM.jl), which supposedly does that, but the package is not documented, nor have there been any commits in the last 3-4 years

Thanks

**Comment by Roger-luo** [comment]
> Is it stored efficiently?

no, none of the `ArrayReg` stores the state in sparse by default, this is because this doesn't matter in statevector emulator.

> I get a vector whose size is 2^num_qubits rather than the number of states

I'm not sure what you meant here, what do you expect to get from it instead?


> I saw many references to active and inactive qubits in Yao, but I did not find an explanation for what that means

active/inactive qubits is just give you a way to apply a (smaller) circuit only partially on the current active qubits, e.g you can `focus!(reg, (1, 3, 5))` then apply a 3-qubit gate on this register, where this gate will act on the 1,3,5-th qubit.

> I used matblock to create a new unitary. is it automatically "reversed" with the adjoint ' operator

yes, you can verify this yourself.

```julia
julia> mat(A)
2×2 Matrix{Float64}:
 0.255424  0.356126
 0.702261  0.082124

julia> mat(A')
┌ Warning: converting Float64 to eltype ComplexF64, consider create another matblock with eltype ComplexF64
└ @ YaoBlocks ~/.julia/packages/YaoBlocks/wUHRm/src/primitive/general_matrix_gate.jl:75
2×2 adjoint(::Matrix{ComplexF64}) with eltype ComplexF64:
 0.255424-0.0im  0.702261-0.0im
 0.356126-0.0im  0.082124-0.0im
```

> I would like to export and import circuits in QASM format. Does Yao support that? I found a repository, [YaoQasm](https://github.com/QuantumBFS/YaoQASM.jl), which supposedly does that, but the package is not documented, nor have there been any commits in the last 3-4 years

It's not actively maintained, but try this package https://github.com/QuantumBFS/YaoBlocksQASM.jl, there are some examples in the tests

**Reply by hros** [reply]
thanks!
> > I get a vector whose size is 2^num_qubits rather than the number of states
> I'm not sure what you meant here, what do you expect to get from it instead?
I was hoping to get a list of states and coefficients. namely, for an `arrayreg` of: $0.6/\sqrt{2}|100001101\rangle+0.8/\sqrt{2}|100010110\rangle+1/\sqrt{2}|101101001\rangle$, I would like to "pseudo measure" the three states and their coefficients and not a vector of 512 elements (mostly zero).

Instead, I have coded the algorithm using arrays of bit strings (instead of states) and an array of coefficients.

thanks for the link to the newer qasm package.  Do you have a list of the current coverage of the project, i.e., what portion of qasm (I guess openqasm 2.0) is covered by the project?

**Reply by Roger-luo** [reply]
The entire OpenQASM 2.0 should be covered and you should be able to convert QASM2.0 compatible Yao circuit to QASM. But there can be bugs due to the lack of maintenance so please report them if you hit one.
